File: ga/glm4_result_2024_6/result/1968D_glm/8_1968D.c
use std::io::{self, Read};

fn min(a: i32, b: i32) -> i32 {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();

    let lines: Vec<&str> = input.lines().collect();

    let t: i32 = lines[0].parse().unwrap();

    for i in 1..=t {
        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;
        let mut a = 0;
        let mut b = 0;

        let mut arr: Vec<i32> = Vec::new();
        let mut aa: Vec<i32> = Vec::new();

        for (j, line) in lines[i * 5..(i * 5) + 5].iter().enumerate() {
            match j {
                0 => {
                    let mut iter = line.split_whitespace();
                    let n: i32 = iter.next().unwrap().parse().unwrap();
                    let k: i32 = iter.next().unwrap().parse().unwrap();
                    a = iter.next().unwrap().parse().unwrap();
                    b = iter.next().unwrap().parse().unwrap();
                    arr.resize(n as usize, 0);
                    aa.resize(n as usize, 0);
                }
                1 => {
                    let nums: Vec<i32> = line.split_whitespace().map(|x| x.parse().unwrap()).collect();
                    arr.copy_from_slice(&nums);
                }
                2 => {
                    let nums: Vec<i32> = line.split_whitespace().map(|x| x.parse().unwrap()).collect();
                    aa.copy_from_slice(&nums);
                }
                _ => {}
            }
        }

        for x in 1..=min(arr.len() as i32, a) {
            p1 = maxl(p1, (aa[a as usize] as i64) * (k - x + 1) as i64 + p1_);
            p1_ += aa[a as usize];
            a = arr[a as usize];
        }

        for x in 1..=min(arr.len() as i32, b) {
            p2 = maxl(p2, (aa[b as usize] as i64) * (k - x + 1) as i64 + p2_);
            p2_ += aa[b as usize];
            b = arr[b as usize];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_mini/8_1968D.rs
use std::io::{self, BufRead};

fn min(a: usize, b: usize) -> usize {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();
    
    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let line = lines.next().unwrap().unwrap();
        let values: Vec<usize> = line.split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        let (n, k, mut a, mut b) = (values[0], values[1], values[2], values[3]);
        let m = min(n, k);

        let mut arr: Vec<usize> = vec![0; n + 1];
        let mut aa: Vec<i64> = vec![0; n + 1];

        let arr_line = lines.next().unwrap().unwrap();
        for (i, val) in arr_line.split_whitespace().enumerate() {
            arr[i + 1] = val.parse().unwrap();
        }

        let aa_line = lines.next().unwrap().unwrap();
        for (i, val) in aa_line.split_whitespace().enumerate() {
            aa[i + 1] = val.parse().unwrap();
        }

        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;

        for x in 1..=m {
            p1 = maxl(p1, aa[a] * (k - x as i64 + 1) + p1_);
            p1_ += aa[a];
            a = arr[a];

            p2 = maxl(p2, aa[b] * (k - x as i64 + 1) + p2_);
            p2_ += aa[b];
            b = arr[b];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}